File: src/classes/join/scan_res.rs
use super::scan_res_item::ScanResItem;
use crate::consts::{errors::HbrsError, response_item::ResponseItem::{DataSize, MinRSSI}};

#[derive(Debug, Clone)]
pub struct ScanRes{
    pub wifis: Vec<ScanResItem>,
}

impl ScanRes{
    pub fn new() -> Self{
        Self{
            wifis: Vec::new(),
        }
    }

    pub fn add_new_item(&mut self, data: Vec<u8>, header_offset: usize) -> Result<bool, HbrsError>{
        if data.len() != DataSize.get_value(){
            return Err(HbrsError::InvalidScanDataLength)
        }

        let new_item = match ScanResItem::new(data, header_offset){
            Ok(item) => item,
            Err(err) => return Err(err),
        };

        // Checks validity of SSID
        if new_item.ssid.iter().all(|&val| val == 0) 
            || new_item.get_ssid_as_str().is_err() 
            || new_item.rssi <= MinRSSI.get_value() as u8{
            return Err(HbrsError::BadSSID);
        }

        // If the item already exists, update the RSSI value
        let existing_item = 
            self.wifis.iter_mut()
                .find( |i|
                i.get_ssid_as_str().unwrap().to_lowercase() 
                == new_item.get_ssid_as_str().unwrap().to_lowercase());
        
        match existing_item{
            Some(existing_item) => {
                if existing_item.rssi < new_item.rssi{
                    existing_item.rssi = new_item.rssi;
                }
            },
            None => {},
        }

        self.wifis.push(new_item);

        return Ok(true);
    }

    pub fn get_bytes(&self) -> Vec<u8>{
        let mut bytes = vec![];

        for wifi in &self.wifis{
            bytes.extend_from_slice(&wifi.get_bytes());
        }

        return bytes;
    }

    pub fn get_wifis(&self) -> Vec<ScanResItem>{
        return self.wifis.clone();
    }

    pub fn set_wifis(&mut self, wifis: Vec<ScanResItem>){
        self.wifis = wifis;
    }

    pub fn get_wifi(&self, ssid: &str) -> Option<ScanResItem>{
        return self.wifis.iter().find(|i| i.get_ssid_as_str().unwrap().to_lowercase() == ssid.to_lowercase()).cloned();
    }

    pub fn get_wifi_by_index(&self, index: usize) -> Option<ScanResItem>{
        return self.wifis.get(index).cloned();
    }

    pub fn get_wifi_count(&self) -> usize{
        return self.wifis.len();
    }
}

File: lib/classes/join/scan_res.py
from lib.classes.join.scan_res_item import ScanResItem
from lib.consts.errors import HbrsError


class ScanRes:
    def __init__(self):
        self.wifis = []

    def add_new_item(self, data, header_offset):
        new_item = ScanResItem(data, header_offset)
        self.wifis.append(new_item)
        """
        if new_item.get_ssid() == b'\x00' * 32 or new_item.get_ssid().decode('utf-8').strip() == '' or new_item.get_rssi() <= 0:
            raise HbrsError.BadSSID

        existing_item = next((i for i in self.wifis if i.get_ssid().decode('utf-8').strip().lower() == new_item.get_ssid().decode('utf-8').strip().lower()), None)
        if existing_item:
            if existing_item.get_rssi() < new_item.get_rssi():
                existing_item.set_rssi(new_item.get_rssi())
        else:
            self.wifis.append(new_item)
        return True
        """
        
        
    def get_bytes(self):
        bytes = b''
        for wifi in self.wifis:
            bytes += wifi.get_bytes()
        return bytes

    def get_wifi(self, ssid):
        return next((i for i in self.wifis if i.get_ssid().decode('utf-8').strip().lower() == ssid.strip().lower()), None)

    def get_wifi_by_index(self, index):
        return self.wifis[index]

    def get_wifis(self):
        return self.wifis